java-checks/src/main/java/org/sonar/java/checks/NullPointerCheck.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012 SonarSource
+ * dev@sonar.codehaus.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.checks;
+
+import org.sonar.api.rule.RuleKey;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.model.expression.MethodInvocationTreeImpl;
+import org.sonar.java.resolve.AnnotationInstance;
+import org.sonar.java.resolve.SemanticModel;
+import org.sonar.java.resolve.Symbol;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.ArrayAccessExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import javax.annotation.Nullable;
+
+@Rule(
+  key = "S2259",
+  name = "Null pointers should not be dereferenced",
+  tags = {"bug", "cert", "cwe", "owasp-a1", "owasp-a2", "owasp-a6", "security"},
+  priority = Priority.BLOCKER)
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.ERRORS)
+@SqaleConstantRemediation("20min")
+public class NullPointerCheck implements JavaFileScanner {
+
+  public static final String KEY = "S2259";
+  private static final RuleKey RULE_KEY = RuleKey.of(CheckList.REPOSITORY_KEY, KEY);
+
+  private JavaFileScannerContext context;
+  private ExpressionVisitor expressionVisitor = new ExpressionVisitor();
+  private IssueVisitor issueVisitor = new IssueVisitor();
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    if (context.getSemanticModel() != null) {
+      expressionVisitor.scan(context.getTree());
+    }
+  }
+
+  /** this visitor invokes issueVisitor for the base expressions of array access, member selectm and method invocation expressions. */
+  private class ExpressionVisitor extends BaseTreeVisitor {
```

Comment:
Why is that  a private class rather than implemented directly by the NullPointerCheck class ?